src/h5web/explorer/utils.ts

```diff
@@ -9,108 +8,83 @@ import {
   MyHDF5Dataset,
   MyHDF5Datatype,
   HDF5HardLink,
-  MyHDF5EntityKind,
   MyHDF5Metadata,
 } from '../providers/models';
+import {
+  makeMyDataset,
+  makeMyDatatype,
+  makeMyGroup,
+  makeMyLink,
+} from '../providers/my-utils';
 import { isGroup, isReachable } from '../providers/utils';
 
-function prepareEntity(
-  link: HDF5HardLink | HDF5RootLink,
-  parent?: MyHDF5Group
-): Omit<MyHDF5Group | MyHDF5Dataset | MyHDF5Datatype, 'kind'> {
-  return {
-    uid: nanoid(),
-    id: link.id,
-    name: link.title,
-    parent,
-    attributes: [],
-    rawLink: link,
-  };
-}
-
 function buildDataset(
   metadata: HDF5Metadata,
-  link: HDF5HardLink | HDF5RootLink,
-  parent?: MyHDF5Group
+  link: HDF5HardLink
 ): MyHDF5Dataset {
-  const rawDataset = metadata.datasets![link.id]; // eslint-disable-line @typescript-eslint/no-non-null-assertion
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const { shape, type, attributes } = metadata.datasets![link.id];
 
-  return {
-    ...prepareEntity(link, parent),
-    kind: MyHDF5EntityKind.Dataset,
-    attributes: rawDataset.attributes || [],
-    shape: rawDataset.shape,
-    type: rawDataset.type,
-  };
+  return makeMyDataset(link.title, shape, type, {
+    id: link.id,
+    attributes,
+    rawLink: link,
+  });
 }
 
 function buildDatatype(
   metadata: HDF5Metadata,
-  link: HDF5HardLink | HDF5RootLink,
-  parent?: MyHDF5Group
+  link: HDF5HardLink
 ): MyHDF5Datatype {
-  const rawDatatype = metadata.datatypes![link.id]; // eslint-disable-line @typescript-eslint/no-non-null-assertion
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const { type } = metadata.datatypes![link.id];
 
-  return {
-    ...prepareEntity(link, parent),
-    kind: MyHDF5EntityKind.Datatype,
-    type: rawDatatype.type,
-  };
+  return makeMyDatatype(link.title, type, {
+    id: link.id,
+    rawLink: link,
+  });
 }
 
 function buildGroup(
   metadata: HDF5Metadata,
-  link: HDF5HardLink | HDF5RootLink,
-  parent?: MyHDF5Group
+  link: HDF5HardLink | HDF5RootLink
 ): MyHDF5Group {
   const rawGroup = metadata.groups[link.id];
 
-  const group: MyHDF5Group = {
-    ...prepareEntity(link, parent),
-    kind: MyHDF5EntityKind.Group,
-    attributes: rawGroup.attributes || [],
-    children: [],
-  };
-
-  group.children = (rawGroup.links || []).map((link) => {
+  const children = (rawGroup.links || []).map((link) => {
     if (!isReachable(link)) {
-      return {
-        uid: nanoid(),
-        name: link.title,
-        kind: MyHDF5EntityKind.Link,
-        parent: group,
-        attributes: [],
-        rawLink: link,
-      };
+      return makeMyLink(link);
     }
```

Comment:
Note that I now initialise the children _before_ the group, as per the "top-down"/"bottom-up" discussion.